File: util/math/src/main/java/dev/sweety/math/pool/Release.java
package dev.sweety.math.pool;

import java.lang.annotation.*;

/**
 * Marks a method that returns a pooled object to its pool, invalidating the reference.
 * After this call the caller must not read from or write to the object.
 *
 * <pre>{@code
 * buf.write(...);
 * buf.release(); // @Release — do not use buf after this line
 * }</pre>
 */
@Documented
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface Release {
}

File: util/math/src/main/kotlin/dev/sweety/math/pool/Release.kt
package dev.sweety.math.pool

/** Marks a method that returns a resource to its pool; the reference must not be used afterwards.  */
@MustBeDocumented
@Retention(AnnotationRetention.BINARY)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class Release 
